### PR TITLE
Fix bug the Enable Banking import.

### PR DESCRIPTION
**Pull request by bernatnan**
Update the origin lines conciliation.

Task: #161817

### CODE REVIEW

File: locale/ca.po [hunk 1]
```diff
@@ -12,12 +12,16 @@ msgstr "Nom ASPSP"
 
 msgctxt "field:account.statement.journal,synchronize_journal:"
 msgid "Synchronize Journal"
-msgstr ""
+msgstr "Diari sincronització"
```
Comment (albertca) [inline locale/ca.po:15]: Si aquesta és la traducció correcta, llavors el títol en anglès és incorrecte i hauria de ser "Synchronization Journal", no?

File: locale/ca.po [hunk 2]
```diff
@@ -12,12 +12,16 @@ msgstr "Nom ASPSP"
...
 
 msgctxt "field:account.statement.origin,entry_reference:"
 msgid "Entry Reference"
 msgstr "Referència d'entrada"
 
+msgctxt "field:account.statement.origin,reconciled:"
+msgid "Reconciled"
+msgstr "Reconciliat"
```
Comment (albertca) [inline locale/ca.po:23]: Reconciliat -> Conciliat

File: locale/ca.po [hunk 3]
```diff
@@ -266,6 +278,22 @@ msgctxt "model:ir.message,text:msg_no_session"
 msgid "Session not found."
 msgstr "Sessió no trobada."
 
+msgctxt "model:ir.message,text:msg_origin_line_with_move"
+msgid ""
+"The line have a move %(move)s related. Delete the line will create a cancell"
```
Comment (albertca) [inline locale/ca.po:283]: Entenc que hauria de ser:

The line has move "%(move)s" related to it. Deleting the line will create a cancelled move. Do you want to continue?

Tot i que no ho acabo d'entendre.

File: locale/ca.po [hunk 4]
```diff
@@ -266,6 +278,22 @@ msgctxt "model:ir.message,text:msg_no_session"
...
+" move. Do you want to continue?"
+msgstr ""
+"La línia té un assentament %(move)s relacionat. Suprimir la línia crearà un "
+"assentament de cancel·lació. Vols continuar?"
+
+msgctxt "model:ir.message,text:msg_origin_pending_amount"
+msgid ""
+"The origin could not be conciled becasue the lines amount is "
```
Comment (albertca) [inline locale/ca.po:291]: conciled -> reconciled
becasue -> because

The origin could not be reconciled because the amount of the lines "%(line_amount)s" and the amount of the origin "%(origin_amount)s" differ.

File: statement.py [hunk 5]
```diff
@@ -4,43 +4,272 @@
...
+        return "%s - %s" % (self.statement.rec_name, self.id)
+
+    @classmethod
+    def search_rec_name(cls, name, clause):
+        return [('statement.rec_name',) + tuple(clause[1:])]
+
+    def get_reconciled(self, name=None):
+        if self.lines and all([True if l.move else False for l in self.lines]):
+            return True
+        return False
```
Comment (albertca) [inline statement.py:84]: Si el que pretens aquí és implementar el camp "conciliat" com el tenim a l'account_bank_statement no és la forma correcta.

El que cal fer és mirar que l'import de les línies coincideixi amb l'import de l'origen. Res a veure amb si té assentament o no té assentament.

Comment (albertca) [reply statement.py:84]: Si el que pretens és saber si s'ha comptabilitzat o no, tampoc és la forma adequada perquè segur que el que volem és poder filtrar i això no és senzill així. Cal afegir un camp "state" a nivell d'origen i fer canvi d'estat al prémer el botó.

File: statement.py [hunk 6]
```diff
@@ -4,43 +4,272 @@
...
+        Statement.write(statements, {
+                'state': 'validated',
+                })
+
     @classmethod
     @ModelView.button
     def reconcile(cls, origins):
         pool = Pool()
-        Line = pool.get('account.statement.line')
+        Statement = pool.get('account.statement')
```
Comment (albertca) [inline statement.py:195]: We need a comment here referencing that it is basically a copy & paste from `account_statement` `create_move()`.

Also I think we must rename this method `create_move()`.

And we need to add a state field to origin and call `create_move()` when moving to `posted` state.

File: statement.py [hunk 7]
```diff
@@ -4,43 +4,272 @@
...
+                self.amount, self.statement.journal.currency)
+            amount = lang.currency(amount, self.statement.journal.currency)
+            raise StatementValidateError(
+                gettext('account_statement_enable_banking.'
+                    'msg_origin_pending_amount',
+                    origin_amount=total_amount,
+                    line_amount=amount))
+
+    @classmethod
+    def validate_statement(cls, statements):
```
Comment (albertca) [inline statement.py:164]: Why naming this method `validate_statement`? Do we really want to validate the whole statement or simply the origins is enough?

I think it should just be named `validate_origin()` and accept origin objects as parameter.

Also I think it must have a comment indicating that it is basically code copy & pasted from account_statement but adapted to work at origin level.

File: statement.py [hunk 8]
```diff
@@ -4,43 +4,272 @@
...
+                raise StatementValidateWarning(warning_key,
+                    gettext('account_statement'
+                        '.msg_statement_invoice_paid_cancelled'))
+            Line.write(paid_cancelled_invoice_lines, {
+                    'related_to': None,
+                    })
+
+        Statement.write(statements, {
+                'state': 'validated',
+                })
```
Comment (albertca) [inline statement.py:189]: We cannot do this.

File: locale/ca.po (hunk 1 above)
Comment (bernatnan) [reply locale/ca.po:15]: Resolt. Em vaig refiar de la traducció automàtica. :-|

File: locale/ca.po (hunk 3 above)
Comment (bernatnan) [reply locale/ca.po:283]: És un avís que salta quan es vol eliminar una línia d'extracte. Avisa que no s'eliminarà l'assentament associat, sinó que es crearà una contrapartida, un assentament de cancel·lació d'aquest. Així mantenim l'històric.

File: statement.py (hunk 5 above)
Comment (bernatnan) [reply statement.py:84]: Volia evitar el camp state per no fer-ho més gros. I simplement tenir un control que si l'origen ja s'havia clicat el botó "Concilia", que aquest ja no aparegués.

File: statement.py (hunk 6 above)
Comment (bernatnan) [reply statement.py:195]: > Also I think we must rename this method `create_move()`.

The name of the function is the same as the button, if we change the function name, we have to change the button and the translation. For me is better the "reconcile" name.

Comment (bernatnan) [reply statement.py:195]: > And we need to add a state field to origin and call `create_move()` when moving to `posted` state.

As I say a few lines above, I wanted to avoid creating a state. But I agree that will be the better solution.

File: statement.py (hunk 7 above)
Comment (bernatnan) [reply statement.py:164]: We need to validate the whole statement, so the name for me is correct.

If we work throw the origins, the statement will always be in draft state. So when we reconcile an origin, we check if its statement has all the origins reconciled and if it's true, we close the statement following the core steps.

File: statement.py (hunk 8 above)
Comment (bernatnan) [reply statement.py:189]: Why?

File: statement.py [hunk 9]
```diff
@@ -4,43 +4,396 @@
...
+                    #to_unpay = [MoveLine(x.id) for x in to_unpay]
+                    Invoice.remove_payment_lines(to_unpay)
+
+                cancel_move = move.cancel()
+                cancel_move.origin = line.origin
+                Move.post([cancel_move])
+                mlines = [l for m in [move, cancel_move]
+                    for l in m.lines if l.account.reconcile]
+                if mlines:
+                    MoveLine.reconcile(mlines)
```
Comment (albertca) [inline statement.py:77]: This method looks a bit naive to me.

I think that at least it should be grouped by account.

File: statement.py [hunk 10]
```diff
@@ -4,43 +4,396 @@
...
+        StatementLine = pool.get('account.statement.line')
+        Move = pool.get('account.move')
+        MoveLine = pool.get('account.move.line')
+
+        cls.validate_origin(origins)
+        cls.create_moves(origins)
+
+        lines = [l for o in origins for l in o.lines]
+        # It's an awfull hack to sate the state, but it's needed to ensure the
+        # Warning of statement state in Move.post is not applied when try to
```
Comment (albertca) [inline statement.py:329]: I guess it's not a Warning but an Error.

File: statement.py [hunk 11]
```diff
@@ -4,43 +4,396 @@
...
+        StatementLine.post_move(lines)
+        if statement_state:
+            Statement.write(*statement_state)
+        # End awful hack
+
+        statements = []
+        for origin in origins:
+            statement = origin.statement
+            try:
+                getattr(statement, 'validate_%s' % statement.validation)()
```
Comment (albertca) [inline statement.py:349]: I think it would be cleaner to create a new validation type that would be the one used in enable banking statements. So no need for this kind of validation.

File: statement.py [hunk 12]
```diff
@@ -4,43 +4,396 @@
...
+                            ))
+                for mline in line.move.lines:
+                    if mline.origin == line:
+                        mline.origin = line.origin
+                        mlines.append(mline)
+                moves.append(line.move)
+        if mlines:
+            with Transaction().set_context(from_account_statement_origin=True):
+                #line_moves = list(set([l.move for l in mlines]))
+                #Move.write(line_moves, {'state': 'draft'})
```
Comment (albertca) [inline statement.py:110]: The commented code should be removed.

File: account.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    __name__ = 'account.move.line'
+
+    @classmethod
+    def _get_origin(cls):
+        return super()._get_origin() + ['account.statement.origin']
+
+    @classmethod
+    def check_modify(cls, *args, **kwargs):
+        context = Transaction().context
+        if Transaction().context.get('from_account_statement_origin',
```
Comment (albertca) [inline account.py:25]: Why is this needed? It will at least require a comment.

File: statement.py (hunk 9 above)
Comment (bernatnan) [reply statement.py:77]: I think I'm not understanding what you mean. Here we take the lines to reconcile to no send all lines, only which the account has the field setted.

File: statement.py (hunk 10 above)
Comment (bernatnan) [reply statement.py:329]: Yes, is an error. Changed the text

File: statement.py (hunk 11 above)
Comment (bernatnan) [reply statement.py:349]: As we discussed, I've tried to do the minimum of new code. But I agree that will be better to have an origin validation type.

File: statement.py (hunk 12 above)
Comment (bernatnan) [reply statement.py:110]: Done.

File: account.py (hunk 13 above)
Comment (bernatnan) [reply account.py:25]: We need to modify the lines even if the move is in 'posted' state.
Add the comment.

File: statement.py (hunk 11 above)
Comment (bernatnan) [reply statement.py:349]: I think is better to use the core validations to ensure the account statements are correct to 'post' them.
